rocprofiler-systems/Jacobi/README.md

```diff
@@ -0,0 +1,228 @@
+
+**In many cases, simply checking the wall_clock files might be sufficient for your profiling!**
+
+If it is not, continue by visualizing the trace.
+
+## Visualizing traces using `Perfetto`
+
+Copy generated `perfetto-trace-0.proto` file to your local machine, and using the Chrome browser open the web page [https://ui.perfetto.dev/](https://ui.perfetto.dev/):
+
+Click `Open trace file` and select the `perfetto-trace-0.proto` file. Below, you can see an example of how the trace file would be visualized on `Perfetto`:
+
+![jacobi_hip-perfetto_screenshot](https://hackmd.io/_uploads/BkgSH-E0A.png)
+
+
+If there is an error opening trace file, try using an older `Perfetto` version, e.g., by opening the web page [https://ui.perfetto.dev/v46.0-35b3d9845/#!/](https://ui.perfetto.dev/v46.0-35b3d9845/#!/).
+
+## Additional features
+### Flat profiles
+
+Append advanced option `ROCPROFSYS_FLAT_PROFILE=true` to `~/.rocprofsys.cfg` or prepend it to the `mpirun` command:
+
+```
+ROCPROFSYS_FLAT_PROFILE=true mpirun -np 1 rocprof-sys-run -- ./Jacobi_hip.inst -g 1 1
+```
+
+`wall_clock-0.txt` file now shows overall time in seconds for each function.
+
+Note the significant total execution time for `hipMemcpy` and `Jacobi_t::Run` calls.
+
+### Hardware counters
+
+To see a list of all the counters for all the devices on the node, do:
+
+```
+rocprof-sys-avail --all
+```
+
+Select the counter you are interested in, and then declare them in your configuration file (or prepend to your `mpirun` command):
+
+```
+ROCPROFSYS_ROCM_EVENTS = VALUUtilization,FetchSize
+```
+
+Run the instrumented binary, and you will observe an output file for each hardware counter specified. You should also see a row for each hardware counter in the `Perfetto` trace generated by rocprof-sys.
+
+Note that you do not have to instrument again after making changes to the config file. Just running the instrumented binary picks up the changes.
+
+```
+ROCPROFSYS_ROCM_EVENTS=VALUUtilization,FetchSize mpirun -np 1 rocprof-sys-run -- ./Jacobi_hip.inst -g 1 1
+cat rocprof-sys-Jacobi_hip.inst-output/<TIMESTAMP>/rocprof-device-0-VALUUtilization-0.txt
+```
+
+### Sampling
+
+To reduce the overhead of profiling, one can use sampling. Set the following in your configuration file (or prepend to your `mpirun` command):
+
+```
+ROCPROFSYS_USE_SAMPLING = true
+ROCPROFSYS_SAMPLING_FREQ = 100
+```
+
+Execute the instrumented binary, inspect `sampling*` files and visualize the `Perfetto` trace:
+
+```
+mpirun -np 1 rocprof-sys-run -- ./Jacobi_hip.inst -g 1 1
+ls rocprofsys-Jacobi_hip.inst-output/<TIMESTAMP>/* | grep sampling
+```
+
+### Profiling multiple MPI processes
+
+Run the instrumented binary with multiple MPI ranks. Note separate output files for each rank, including `perfetto-trace-*.proto` and `wall_clock-*.txt` files.
+
+```
+mpirun -np 2 rocprof-sys-run -- ./Jacobi_hip.inst -g 2 1
+```
+
+Inspect output text files. Then visualize `perfetto-trace-*.proto` files in `Perfetto`.
```

Comment:
It will be nice to add a section here about how to concatenate the various proto files and view the merged profile in Perfetto. A command such as `cat perfetto-trace-*.proto > merged.proto` would be sufficient for merging them.